tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -281,86 +303,109 @@ bpf_code_generator::program_sections()
+    const GUID& ptype, const GUID& attach_type, const std::string& hash_type)
 {
-    memcpy(&current_program->program_type, &program_type, sizeof(GUID));
-    memcpy(&current_program->expected_attach_type, &attach_type, sizeof(GUID));
-    current_program->program_info_hash_type = program_info_hash_type;
+    memcpy(&this->program_type, &ptype, sizeof(GUID));
+    memcpy(&this->expected_attach_type, &attach_type, sizeof(GUID));
+    this->program_info_hash_type = hash_type;
 }
 
 void
-bpf_code_generator::set_program_hash_info(const std::optional<std::vector<uint8_t>>& program_info_hash)
+bpf_code_generator::set_program_hash_info(
+    const unsafe_string& program_name, const std::optional<std::vector<uint8_t>>& program_info_hash)
 {
-    current_program->program_info_hash = program_info_hash;
+    bpf_code_generator_program& program = programs[program_name];
+
+    program.program_info_hash = program_info_hash;
 }
 
 void
-bpf_code_generator::generate(
-    const bpf_code_generator::unsafe_string& section_name, const bpf_code_generator::unsafe_string& program_name)
+bpf_code_generator::generate(const bpf_code_generator::unsafe_string& program_name)
 {
-    current_program = &programs[program_name];
+    bpf_code_generator_program& program = programs[program_name];
 
-    generate_labels();
-    build_function_table();
-    encode_instructions(section_name);
+    program.generate_labels();
+    program.build_function_table();
+    program.encode_instructions(map_definitions);
 }
 
 std::vector<int32_t>
-bpf_code_generator::get_helper_ids()
+bpf_code_generator::get_helper_ids(const bpf_code_generator::unsafe_string& program_name)
 {
+    bpf_code_generator_program& program = programs[program_name];
     std::vector<int32_t> helper_ids;
-    for (const auto& [name, helper] : current_program->helper_functions) {
+    for (const auto& [name, helper] : program.helper_functions) {
         helper_ids.push_back(helper.id);
     }
 
     return helper_ids;
 }
 
 void
-bpf_code_generator::extract_program(const ebpf_api_program_info_t* program_info)
+bpf_code_generator::extract_program(
+    const ebpf_api_program_info_t* program_info, _In_ const ebpf_api_program_info_t* infos)
 {
-    std::vector<ebpf_inst> program{
+    std::vector<ebpf_inst> instructions{
         reinterpret_cast<const ebpf_inst*>(program_info->raw_data),
         reinterpret_cast<const ebpf_inst*>(program_info->raw_data + program_info->raw_data_size)};
 
+    bpf_code_generator_program& program = programs[program_info->program_name];
+
     uint32_t offset = 0;
-    for (const auto& instruction : program) {
-        current_program->output.push_back({instruction, offset++});
+    size_t end_index = instructions.size();
+    for (size_t index = 0; index < end_index; index++) {
+        const auto& instruction = instructions[index];
+
+        // At this point, instructions will contain all subprograms appended.
```

Comment:
Is appending the sub-programs something that loader does? I thought each subprogram was in it's own section.

Reply:
Expanded comment explaining.